source/opt/loop_descriptor.h

```diff
@@ -0,0 +1,262 @@
+
+  // Returns the loop pre-header, nullptr means that the loop predecessor does
+  // not qualify as a preheader.
+  // The preheader is the unique predecessor that:
+  //   - Dominates the loop header;
+  //   - Has only the loop header as successor.
+  inline BasicBlock* GetPreHeaderBlock() { return loop_preheader_; }
+
+  // Returns the loop pre-header.
+  inline const BasicBlock* GetPreHeaderBlock() const { return loop_preheader_; }
+
+  // Returns true if this loop contains any nested loops.
+  inline bool HasNestedLoops() const { return nested_loops_.size() != 0; }
+
+  // Returns the depth of this loop in the loop nest.
+  // The outer-most loop has a depth of 1.
+  inline size_t GetDepth() const {
+    size_t lvl = 1;
+    for (const Loop* loop = GetParent(); loop; loop = loop->GetParent()) lvl++;
+    return lvl;
+  }
+
+  // Adds |nested| as a nested loop of this loop. Automatically register |this|
+  // as the parent of |nested|.
+  inline void AddNestedLoop(Loop* nested) {
+    assert(!nested->GetParent() && "The loop has another parent.");
+    nested_loops_.push_back(nested);
+    nested->SetParent(this);
+  }
+
+  inline Loop* GetParent() { return parent_; }
+  inline const Loop* GetParent() const { return parent_; }
+
+  inline bool HasParent() const { return parent_; }
+
+  // Returns true if this loop is itself nested within another loop.
+  inline bool IsNested() const { return parent_ != nullptr; }
+
+  // Returns the set of all basic blocks contained within the loop. Will be all
+  // BasicBlocks dominated by the header which are not also dominated by the
+  // loop merge block.
+  inline const BasicBlockListTy& GetBlocks() const {
+    return loop_basic_blocks_;
+  }
+
+  // Returns true if the basic block |bb| is inside this loop.
+  inline bool IsInsideLoop(const BasicBlock* bb) const {
+    return IsInsideLoop(bb->id());
+  }
+
+  // Returns true if the basic block id |bb_id| is inside this loop.
+  inline bool IsInsideLoop(uint32_t bb_id) const {
+    return loop_basic_blocks_.count(bb_id);
+  }
+
+  // Adds the Basic Block |bb| this loop and its parents.
+  void AddBasicBlockToLoop(const BasicBlock* bb) {
+#ifndef NDEBUG
+    assert(bb->GetParent() && "The basic block does not belong to a function");
+    IRContext* context = bb->GetParent()->GetParent()->context();
+
+    opt::DominatorAnalysis* dom_analysis =
+        context->GetDominatorAnalysis(bb->GetParent(), *context->cfg());
+    assert(dom_analysis->Dominates(GetHeaderBlock(), bb));
+
+    opt::PostDominatorAnalysis* postdom_analysis =
+        context->GetPostDominatorAnalysis(bb->GetParent(), *context->cfg());
+    assert(postdom_analysis->Dominates(GetMergeBlock(), bb));
+#endif  // NDEBUG
+
+    for (Loop* loop = this; loop != nullptr; loop = loop->parent_) {
+      loop_basic_blocks_.insert(bb->id());
+    }
+  }
+
+  // Returns true if the parent basic block of |inst| belong to this loop.
+  inline bool IsLoopInvariant(Instruction* inst) const {
```

Comment:
This is confusing to me.  My understanding of a loop invariant instruction is one that computes the same value for every iteration of the loop.  This seems to be true if the instruction is inside the loop, but it says nothing about the value being computed.

Reply:
Well I totally agree with you, I suppose a better name for the function would be more "IsDynamicallyLoopInvariant" (are we recomputing this value at each step). But after a loop invariant code motion, this also say something about the value being computed.

That's implemented this way for performance reasons. To take into account the value being computed, we need to walk through the use-def chain for each query and so to make this efficient we will need to cache the result. The resulting problem is that the cache could quickly become invalid and selective invalidation might be intrusive for users. I have the feeling that this would be more efficient (in terms of memory and speed) if it stay this way.

Reply:
To avoid confusion, can we rename the function to overload "IsInsideLoop"?  It is a much easier to understand and does not carry with the understanding of "invariance", which is really a statement of the value computed.